### DISCUSSION TITLE
Server on a network for multiple users

### DISCUSSION THREAD

**Discussion by Spinaster** [Comfy-Org/ComfyUI#4185] [General]
I have implemented a ComfyUI server on a Mac machine. Currently, users connect to the same address and use the server simultaneously. The problem is that all the process (image generation) do not occur simultaneously but are queued as if there were a single user. This causes delays.
Is this the correct approach to use a server on a network for multiple users? Is there a solution?
Thanks!

**Comment by ltdrdata** [comment]
Fundamentally, ComfyUI was not designed for multi-user use, and AI image generation utilizes the GPU fully. 
Even if parallelization were allowed, it would ultimately be meaningless because the GPU usage portion cannot be parallelized.
(In workflow execution, most of the time is spent using the GPU.)

**Comment by Lex-DRL** [comment]
... however, a concurrent "parallelism" would be much appreciated.

Here's a use case for a single-houshold or even single-user multi-workflow scenario (i.e., when security/privacy isn't an issue):
- You have two workflows open in different browser tabs (either by `hosts` hack you @ltdrdata menioned [here](https://github.com/comfyanonymous/ComfyUI/issues/1239#issuecomment-1681487932) or simply by opening ComfyUI from different computers).
- Currently, each time you switch between workflows (do something here -> do something there), everything previously generated is lost, and the entire workflow needs to be generated from scratch.
- Even a simple "state saving" would be a HUUUGE improvement. I.e., offloading previously generated graph state from VRAM to RAM (and further to SWAP/TMP storage) when switching between workflows. Then - yes, you'll need to wait while GPU frees/reloads the workflow state from/to VRAM each time it can't fit both there - but it's still at least SOME form of working with multiple workflows in parallel (which, by the way, can share their results to each other via simple save-refresh-load).

Yes, it's a very rudimentary "parallelism", but even it would be a massive improvement in terms of multitasking.

**Comment by bigcat88** [comment]
Do you need this for the parallel development of workflows, or just for using existing ones?

If you only need it for using existing workflows and not for development, there are solutions available that aim to do that and use ComfyUI under the hood, such as [StableSwarmUI](https://github.com/Stability-AI/StableSwarmUI) or [Visionatrix](https://github.com/Visionatrix/Visionatrix)

**Comment by Lex-DRL** [comment]
Thanks for the suggestion. I know of StableSwarm, but it's not a real solution for an underlying problem. It's a workaround at best.

In reality, you almost never "develop workflows" and "use them" as two separate processes *(the latter is an activity for complete beginners and, as a result, for youtube bloggers on the other end - as an effective tool for self-promotion and **TEASING** what can be done with ComfyUI to someone who has never saw any node editor before)*.
Most of the time, the whole process of generating **exactly** what you want is a back and forth of tweaking parameters and changing the workflow itself. Any well-established node editor proves that (Houdini/Nuke/Unreal Blueprints/Unity ShaderGraph/Substance Designer...). The entire content creation there **IS** "workflow development".

And since ComfyUI has no support for subgraphs, at least splitting one enormously giant spaghetti mess into smaller, a bit more manageable messes is kinda the obvious way to go for any kind of complex networks. Until we get proper subgraphs, of course.

So no, I don't think a GUI wrapper like Swarm is a true solution for the issue. The true one is finally getting subgraphs in the core... and I believe getting multiple tabs to work concurrently is a necessary step to get there.

**Comment by Mikec78660** [comment]
So my use case is this. I have two tesla p40 gpus ( cuda:0 and cuda:1 and two tesla p4 gpus (cuda:2 and cuda:3) in my machine. I can make a workflow and have the same workflow open in two browser tabs. I load a lara's on each p4 and the checkpoint model on one of the p40's in one tab and the other p40 in the other tab but it's won't run the second workflow until the first has completed even though the only thing being used in the first p40. Would be great if it would run them in parallel.
